src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -1054,6 +1102,246 @@ get_rebalance_progress(PG_FUNCTION_ARGS)
+	int queryResult = ExecuteOptionalRemoteCommand(connection, query->data, &result);
+	if (queryResult != RESPONSE_OKAY)
+	{
+		ereport(ERROR, (errcode(ERRCODE_CONNECTION_FAILURE),
+						errmsg("cannot get the size because of a connection error")));
+	}
+
+	int rowCount = PQntuples(result);
+	int colCount = PQnfields(result);
+
+	if (colCount < 2)
+	{
+		ereport(ERROR, (errmsg("unexpected number of columns returned by: %s",
+							   query->data)));
+	}
+
+	HASHCTL info = {
+		.keysize = sizeof(uint64),
+		.entrysize = sizeof(ShardPlacementStatistics),
+		.hcxt = CurrentMemoryContext
+	};
+
+	HTAB *shardStatistics = hash_create("ShardPlacementStatistics", 32, &info,
+										HASH_ELEM | HASH_CONTEXT | HASH_BLOBS);
+
+	int rowIndex = 0;
+	for (; rowIndex < rowCount; rowIndex++)
+	{
+		char *shardIdString = PQgetvalue(result, rowIndex, 0);
+
+		uint64 shardId = pg_strtouint64(shardIdString, NULL, 10);
+		uint64 totalSize = 0;
+
+		if (!PQgetisnull(result, rowIndex, 1))
+		{
+			char *sizeString = PQgetvalue(result, rowIndex, 1);
+			totalSize = pg_strtouint64(sizeString, NULL, 10);
+		}
+
+		ShardPlacementStatistics *statistics =
+			hash_search(shardStatistics, &shardId, HASH_ENTER, NULL);
+		statistics->totalSize = totalSize;
+	}
+
+	PQclear(result);
+
+	bool raiseErrors = false;
+	ClearResults(connection, raiseErrors);
+
+	return shardStatistics;
+}
+
+
+static HTAB *
+GetShardMoveShardIdsByWorker(PlacementUpdateEventProgress *steps, int stepCount,
+							 bool fromSource)
+{
+	HASHCTL info = {
+		.keysize = sizeof(WorkerHashKey),
+		.entrysize = sizeof(WorkerShardMoveShardIds),
+		.hcxt = CurrentMemoryContext
+	};
+
+	HTAB *shardsByWorker = hash_create("GetRebalanceStepsByWorker", 32, &info,
+									   HASH_ELEM | HASH_CONTEXT | HASH_BLOBS);
+
+	for (int stepIndex = 0; stepIndex < stepCount; stepIndex++)
+	{
+		PlacementUpdateEventProgress *step = &(steps[stepIndex]);
+
+		AddShardIdToWorkerList(shardsByWorker, step->sourceName, step->sourcePort,
+							   &step->shardId);
+
+		if (step->progress == 0)
+		{
+			/*
+			 * shard move has not started so we don't need target stats for
+			 * this shard
+			 */
+			continue;
+		}
+
+		AddShardIdToWorkerList(shardsByWorker, step->targetName, step->targetPort,
+							   &step->shardId);
+	}
+
+	return shardsByWorker;
+}
+
+
+static void
+AddShardIdToWorkerList(HTAB *shardsByWorker, char *workerName, int workerPort,
+					   uint64 *shardId)
+{
+	WorkerHashKey workerKey;
+	memset_struct_0(workerKey);
+
+	strlcpy(workerKey.hostname, workerName, MAX_NODE_LENGTH);
+	workerKey.port = workerPort;
+
+	bool isFound = false;
+	WorkerShardMoveShardIds *sourceShardIds =
+		hash_search(shardsByWorker, &workerKey, HASH_ENTER, &isFound);
+	if (!isFound)
+	{
+		sourceShardIds->shardIds = NIL;
+	}
+
+	/*
+	 * TODO: is it ok that we are pointing into shared memory here? We're not
+	 * changing it so I expect it should be fine
+	 */
+	sourceShardIds->shardIds = lappend(sourceShardIds->shardIds, shardId);
```

Comment:
Changed this thing to be a hashset instead of a list, which does the copy itself. It's possible that a shard is moved twice, so a set is also more correct. Otherwise we'd get the same shard in the resulting statistics query twice for a single node.